ethereum/mist
# Implement Ethereum Provider (EIP 1193)

### What does it do?

This PR deprecates providing a versioned web3 library on `window.web3` and implements an Ethereum Provider given to dapps on `window.ethereum` as specified in [EIP-1193.md](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md). 

### To test/review

Run wallet in dev on branch [meteor-dapp-wallet/ethereum-provider](https://github.com/ethereum/meteor-dapp-wallet/pull/487) and ensure wallet on http localhost port works correctly.

Browse to a new website, open the console, and try running some commands through `window.ethereum`:

`window.ethereum.send('eth_accounts').then(console.log).catch(console.error);`

This should return `[]` by default on a new tab without any connected accounts.

Try connecting an account or two, perhaps by running:

`window.ethereum.send('eth_requestAccounts').then(console.log).catch(console.error);`

then running again:

`window.ethereum.send('eth_accounts').then(console.log).catch(console.error);`

Or an event with type accountsChanged will emit and if listening you can see the new accounts populate through there. If you close the connect window without connecting any accounts, the dapp should receive a `User Denied Full Provider` error to update next state in UI.

You can also try sending a transaction `from` an account that's not yours, and it will reject with an `Unauthorized` error.

You can also subscribe to new blocks in console:

```js
ethereum
  .send('eth_subscribe', ['newHeads'])
  .then(subscriptionId => {
    subId = subscriptionId;
    ethereum.on('notification', result => {
      if (result.subscription === subscriptionId) {
        if (result.result instanceof Error) {
          const error = result.result;
          console.error(
            `Error from newHeads subscription: ${error.message}.
             Code: ${error.code}. Data: ${error.data}`
          );
        } else {
          const block = result.result;
          console.log(`New block ${block.number}:`, block);
        }
      }
    });
  })
  .catch(error => {
    console.error(
      `Error making newHeads subscription: ${error.message}.
       Code: ${error.code}. Data: ${error.data}`
    );
  });
```

### Full QA checklist

* [ ] Test provider methods
  * [ ] send
  * [ ] on
* [ ] Test provider events
  * [ ] connect
  * [ ] networkChanged
  * [ ] accountsChanged

#### Provider methods
To test these methods, run the commands in the console of a webview.

##### requestAccounts
```js
window.ethereum.send('eth_requestAccounts').then(console.log).catch(console.error);`
```
Should return `true` if connected accounts, otherwise returns [error code](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#error-object-and-codes) 4001.
 
##### send
```js
window.ethereum.send('eth_accounts').then(console.log).catch(console.error);
```
Should return array of connected accounts, or `[]` if none.

```js
ethereum.send('eth_sendTransaction', [{from: '0x00A839dE7922491683f547a67795204763ff8237', to: '0xe6476bd2f10DBe8a4CA2d08dD117A225aa863A23'}]).then(console.log).catch(console.error)
````
Should return `Unauthorized` if using non-connected account. Change to `from` a connected account and it should successfully open the sendTx window.

##### subscribe
```js
var subId;
ethereum
  .send('eth_subscribe', ['newHeads'])
  .then(subscriptionId => {
    subId = subscriptionId;
    ethereum.on('notification', result => {
      if (result.subscription === subscriptionId) {
        if (result.result instanceof Error) {
          const error = result.result;
          console.error(
            `Error from newHeads subscription: ${error.message}.
             Code: ${error.code}. Data: ${error.data}`
          );
        } else {
          const block = result.result;
          console.log(`New block ${block.number}:`, block);
        }
      }
    });
  })
  .catch(error => {
    console.error(
      `Error making newHeads subscription: ${error.message}.
       Code: ${error.code}. Data: ${error.data}`
    );
  });
```
##### unsubscribe
After subscription above starts returning:
```js
ethereum
  .send('eth_unsubscribe', [subId])
  .then(result => {
    if (result) {
      console.log(`Unsubscribed newHeads subscription ${subId}`);
    }
  })
  .catch(error => {
    console.error(
      `Error unsubscribing newHeads subscription: ${error.message}.
       Code: ${error.code}. Data: ${error.data}`
    );
  });
```
#### Provider events
To test these events, run the commands in the console of a webview then try changing networks and changing accounts.
##### connect
```js
ethereum.on('connect', () => {
  console.log(`Ethereum provider connection connected`);
});
```
##### close
```js
ethereum.on('close', (code, reason) => {
  console.log(`Ethereum provider connection closed: ${reason}. Code: ${code}`);
});
```
##### networkChanged
```js
ethereum.on('networkChanged', (networkId) => {
  console.log(`Network changed to new ID: ${networkId}`);
});
```
##### accountsChanged
```js
const logAccounts = accounts => {
  console.log(`Accounts:\n${accounts.join('\n')}`);
};
ethereum.on('accountsChanged', logAccounts);
// to unsubscribe
ethereum.removeListener('accountsChanged', logAccounts);
```
